File: python/day06.py
from aocd import get_data
from dotenv import load_dotenv

load_dotenv()


def day06(message: str, n: int = 4) -> int:
    for i in range(n - 1, len(message)):
        if len(set(message[i-n:i])) == n:
            return i


if __name__ == '__main__':
    data = get_data(day=6, year=2022)
    print("Number of characters processed (part a):", day06(data))
    print("Number of characters processed (part b):", day06(data, n=14))

File: rust/day06.rs
use std::fs;
use std::iter::FromIterator;
use std::collections::HashSet;


fn day06 (data: &str, n: usize) -> u16 {
    let message: Vec<char> = data.chars().collect();
    for (i, s) in message.windows(n as usize).enumerate() {
        let hs: HashSet<&char> = HashSet::from_iter(s);
        let unique_chars: Vec<_> = hs.into_iter().collect();
        if unique_chars.len() == n {
            return (i+n) as u16;
        }
    }
    return 3;
}


fn main() {
    let contents = fs::read_to_string("inputs/day06.txt")
        .expect("Should have been able to read the file");
    println!("Number of characters parsed (part a): {}", day06(&contents, 4));
    println!("Number of characters parsed (part b): {}", day06(&contents, 14));
}